### PR TITLE
week5 - linked list palindrome

### PR DESCRIPTION
**Pull request by maknoon**
- `$ make pLL && ./pLL`
- TEBOW-IT took ~40min. to optimize runtime just a tiny bit, solution stores the first half of the LL in a stack and then compares it with the second half of the LL (ignoring the middle element if the size of LL is odd). I spent a long time trying to figure out how to optimize the space complexity, because I think there is definitely a way to "reverse" the LL and then test it for being a palindrome without using a stack, while still being O(n).
- `TODO`: come back and figure out the other way. also should probably move the `node` struct & those "LL class member" functions to an actual class / include the header file to organize my code better

### CODE REVIEW

File: week5/palindromeLL.cc [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+ */
+bool palindromeLL(node *head)
+{
+  // case: null or only one -> is a palindrome
+  if (!head || (!head->next)) return true;
+
+  // find the middle node of the LL by using a faster 'runner' node
+  // since runner goes 2x as fast as mid, by the time runner is at the end,
+  // the mid node will be in the middle of the LL
+  node *mid = head;
```
Comment (loresfca) [inline week5/palindromeLL.cc:82]: this is a really clever way to find the middle :)

File: week5/palindromeLL.cc [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+/*
+ * bool palindromeLL(node *head)
+ * inputs: *head - pointer to the starting node of the LL
+ * output: T if the LL is a palindrome, F if not
+ * notes: main solution function
+ */
+bool palindromeLL(node *head)
+{
+  // case: null or only one -> is a palindrome
+  if (!head || (!head->next)) return true;
```
Comment (loresfca) [inline week5/palindromeLL.cc]: Just a little detail here, I think that you should return an exception/error when the head is null

**Review by loresfca — COMMENTED** [review]
Great work as always! Your solution is really solid and clear! I think this is the optimal solution without modifying the list, however, I think there is a way to use O(1) space by reversing as you said and that's the only thing that could improve your solution :D

File: week5/palindromeLL.cc [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+  }
+
+  // if we reach this point then we only need to verify the first and last elements
+  return (c->data == palindrome.top());
+}
+
+/*
+ * for testing purposes
+ */
+void tests(void) {
```
Comment (charientism) [inline week5/palindromeLL.cc:109]: Thanks for uploading your written TEBOW IT notes and your test cases! I was going to leave feedback to try to pick simple pass/fail, even/odd cases for your examples, but you did that in the code below :P

File: week5/palindromeLL.cc [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+//== FUNCTIONS
+/*
+ * void insert(node*& head, int key)
+ * inputs: head - the LL node to insert the new node in front of
+           key - value of the new node to be inserted
+ * output: none
+ * notes: helper function to insert a new node into the LL, as the new head node
+ */
+void insert(node*& head, char d)
```
Comment (charientism) [inline week5/palindromeLL.cc:39]: Great comments (as usual :)). I'm also really happy with how you structure your code with helper methods to keep everything clear/clean.

File: week5/palindromeLL.cc (hunk 2 above)
Comment (charientism) [reply week5/palindromeLL.cc]: Definitely something to ask the interviewer about. I saw it in your Talk TEBOW IT notes so I know it was something you considered

File: week5/palindromeLL.cc [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+  while (runner && runner->next) {
+    runner = runner->next->next;
+    mid = mid->next;
+  }
+
+  // use a stack for the first half of LL
+  stack<char> palindrome;
+  node *c;
+  for (c = head; c != mid; c = c->next) palindrome.push(c->data);
+  if (runner) c = c->next; // SKIP the middle element if size of LL is odd
```
Comment (charientism) [inline week5/palindromeLL.cc:93]: Nice job watching out for and handling this case. Interviewers love little got-cha's like this and paying close attention to the details of your implementation pays major dividends

File: week5/palindromeLL.cc [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+  // check remaining half to verify palindrome
+  while (c->next) {
+      if (c->data != palindrome.top()) return false;
+      palindrome.pop();
+      c = c->next;
+  }
+
+  // if we reach this point then we only need to verify the first and last elements
+  return (c->data == palindrome.top());
```
Comment (charientism) [inline week5/palindromeLL.cc:103]: Great job with your week 5 hw. You're correct that you can reverse a linked list with O(1) additional space. It's actually not something that we'll go over tomorrow (in fact, you'll see a solution shockingly similar to yours :)), but it's a nice optimization and a great practice problem.

**Review by charientism — APPROVED** [review]

File: week5/palindromeLL.cc (hunk 2 above)
Comment (maknoon) [reply week5/palindromeLL.cc]: fixed to throw null exception

File: week5/palindromeLL.cc (hunk 4 above)
Comment (maknoon) [reply week5/palindromeLL.cc:39]: thanks! :)

**Pull request merged**